lib/uaa/token_coder.rb

```diff
@@ -115,6 +117,17 @@ def self.decode(token, options = {}, obsolete1 = nil, obsolete2 = nil)
       raise InvalidSignature, "Signature verification failed" unless
           options[:skey] && constant_time_compare(signature, OpenSSL::HMAC.digest(init_digest(algo), options[:skey], signing_input))
     elsif ["RS256", "RS384", "RS512"].include?(algo)
+      if !options[:pkey] && options[:info] && signing_key_name = header["kid"]
+        token_keys = options[:info].cached_validation_keys_hash
+        if token_keys && !token_keys[signing_key_name]
+          # New/unknown kid signing key name, refresh cache from /token_keys
+          token_keys = options[:info].cached_validation_keys_hash(reload: true)
+        end
+        if token_keys && token_keys[signing_key_name] && token_keys[signing_key_name]["value"]
+          verification_key = token_keys[signing_key_name]["value"]
+          options[:pkey] = OpenSSL::PKey::RSA.new(verification_key)
```

Comment:
I'm not a Ruby coder, so take this with a grain of salt, but is it safe to be making assignments to the `options` object? If this is used how I think it is, wouldn't it be common for a user of the library to create a "static" instance of the `options` object during program initialisation, and this be held onto by the client object, which may be serving multiple requests from multiple threads, some of which may have keys with one ID, and others with keys with a newly rotated ID.

In that case there is no guarantee that the value set on this line will be read back again in the subsequent verification step.

To fix, wouldn't we just create a local variable instead to hold this public key?

Reply:
I'm unsure of exact Ruby behavior here either; so new commit will make `pkey` a local variable.